### PR TITLE
Add ability statically link OpenSSL

### PR DESCRIPTION
**Pull request by kant2002**
This setup works if I apply this as local customizations, I do not sure that I use OpenSSL in most secure way, I do not competent. This setup and StaticExecutable=true allow package just EXE file + /etc/ssl/certs/ folder in Docker

### CODE REVIEW

File: src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets [hunk 1]
```diff
@@ -87,6 +95,8 @@ The .NET Foundation licenses this file to you under the MIT license.
 
     <Exec Command="$(IlcHostPackagePath)/native/src/libs/System.Globalization.Native/local_build.sh $(IlcHostPackagePath)/ $(IntermediateOutputPath)" Condition="'$(StaticICULinking)' == 'true'"/>
 
+    <Exec Command="$(IlcHostPackagePath)/native/src/libs/System.Security.Cryptography.Native/local_build.sh $(IlcHostPackagePath)/ $(IntermediateOutputPath)" Condition="'$(StaticOpenSslLinking)' == 'true'"/>
```
Comment (jkotas) [inline src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets:98]: `local_build.sh` file missing in the PR?

Comment (kant2002) [reply src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets:98]: Oops. But it's the same as in Globalization work basically.

Comment (jkotas) [reply src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets:98]: > same as in Globalization

Maybe we should have just one script and pass the library to build as an argument?

**Comment by vcsjones** [comment]
@bartonjs @blowdart a heads up this is proposing a change to allow NativeAOT apps to statically link OpenSSL (`libcrypto` and `libssl`).

**Comment by blowdart** [comment]
This change would mean any Microsoft product using aot would have to get a cryptoboard exception. Static linking openssl is arguably rather bad security. So many products out there in routers, IoT devices etc. cannot even manage to update dynamically linked OpenSSL, a static linkage makes its use even more hidden.

From a compliance and security PoV if reject this. Has there been a wider discussion I've missed, or is it just this PR?

Cc @GrabYourPitchforks

**Comment by kant2002** [comment]
For perspective why I was adding this change - I was thinking to have application as single EXE file running inside docker containers built from scratch image. If update needed, then docker container recreated. I do not consider that this would be used in different contexts.

**Comment by jkotas** [comment]
> This change would mean any Microsoft product using aot would have to get a cryptoboard exception.

@blowdart That's not correct. This change is not changing any default. The aot build is going to dynamically link openssl by default both before and after this change.

This change is adding option to allow statically linking. This is expert-only option. It requires acquiring extra dependencies and it mirrors what we have done for other libraries (ICU). I do not expect that we are going to document this option in official documentation. 

The community has always been interested having an option to statically linking everything together to build "from scratch" container images, etc. This small change supports makes this niche scenario a bit easier.

> Static linking openssl is arguably rather bad security. 

It depends on how you orchestrate your deployment pipeline. If your deployment pipeline fully rebuilds your containers for updates, there is a little difference between dynamic and static linking security-wise. You have to always rebuild your container to pick-up OpenSSL update for both cases. The only difference between these two cases are performance characteristics.

> Has there been a wider discussion I've missed, or is it just this PR?

This is follow up on https://github.com/dotnet/runtime/issues/79498

**Comment by blowdart** [comment]
Oh dear, ok, then any AOT compiled app which opts-in will need a cryptoboard exception, before or after this *sigh*

Does the documentation call this sort of thing out with a large warning/caveat anywhere?

File: src/native/libs/build-local.sh [hunk 2]
```diff
@@ -0,0 +1,27 @@
+#!/bin/sh
+
+# Licensed to the .NET Foundation under one or more agreements.
+# The .NET Foundation licenses this file to you under the MIT license.
+#
+
+# This script is used only for building libSystem.Security.Cryptography.Native.a
```
Comment (jkotas) [inline src/native/libs/build-local.sh:7]: Outdated comment

**Comment by jkotas** [comment]
> Oh dear, ok, then any AOT compiled app which opts-in will need a cryptoboard exception, before or after this sigh

No AOT compiled app shipped by Microsoft should be opting-in into this.

> Does the documentation call this sort of thing out with a large warning/caveat anywhere?

This option is not documented at all today. You have to find in the source. We can certainly add documentation for this with a disclaimer, but it is only going to make the option more discoverable, and I expect that more people will try to use it.

File: src/native/libs/build-local.sh [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+# Currently, only Linux is supported
+
+SHIM_SOURCE_DIR=$1/native/src
+INTERMEDIATE_OUTPUT_PATH=$2
+TARGET_LIBRARY=$3
+
+if [ -d "$SHIM_SOURCE_DIR" ]; then
+    LOCAL_SHIM_DIR="$INTERMEDIATE_OUTPUT_PATH"/libs/$TARGET_LIBRARY/build
+    mkdir -p "$LOCAL_SHIM_DIR" && cd "$LOCAL_SHIM_DIR"
+    if [ $? -ne 0 ]; then echo "build-local.sh::ERROR: Cannot use local build directory"; exit 1; fi
```
Comment (am11) [inline src/native/libs/build-local.sh:20]: Please use the same inline syntax as before for compat. The only change needed was replacing hardcoded name with `$TARGET_LIBRARY`.

**Comment by jkotas** [comment]
@kant2002 Could you please mention this next to https://github.com/dotnet/runtime/blob/main/src/coreclr/nativeaot/docs/compiling.md#using-statically-linked-icu and include a large warning/caveat ?

File: src/native/libs/build-local.sh (hunk 3 above)
Comment (kant2002) [reply src/native/libs/build-local.sh]: Thank you! I copy file from browser instead of file system

File: src/native/libs/build-local.sh [hunk 4]
```diff
@@ -4,24 +4,25 @@
...
 # in the end-user's computer for NativeAOT purposes (static linking).
 # This file is not used during the dotnet runtime build.
 
 # Currently, only Linux is supported
 
-SHIM_SOURCE_DIR="$1"/native/src
-INTERMEDIATE_OUTPUT_PATH="$2"
+SHIM_SOURCE_DIR=$1/native/src
+INTERMEDIATE_OUTPUT_PATH=$2
+TARGET_LIBRARY=$3
```
Comment (am11) [inline src/native/libs/build-local.sh:15]:
```suggestion
SHIM_SOURCE_DIR="$1"/native/src
INTERMEDIATE_OUTPUT_PATH="$2"
TARGET_LIBRARY="$3"
```

For paths with spaces (and to make `shellcheck(1)` happy 🙂).

File: src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets [hunk 5]
```diff
@@ -85,7 +93,9 @@ The .NET Foundation licenses this file to you under the MIT license.
       <NativeFramework Include="GSS" />
     </ItemGroup>
 
-    <Exec Command="$(IlcHostPackagePath)/native/src/libs/System.Globalization.Native/local_build.sh $(IlcHostPackagePath)/ $(IntermediateOutputPath)" Condition="'$(StaticICULinking)' == 'true'"/>
+    <Exec Command="$(IlcHostPackagePath)/native/src/libs/build-local.sh $(IlcHostPackagePath)/ $(IntermediateOutputPath) System.Globalization.Native" Condition="'$(StaticICULinking)' == 'true'"/>
+
+    <Exec Command="$(IlcHostPackagePath)/native/src/libs/build-local.sh $(IlcHostPackagePath)/ $(IntermediateOutputPath) System.Security.Cryptography.Native" Condition="'$(StaticOpenSslLinking)' == 'true'"/>
```
Comment (am11) [inline src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets:98]: nit: quotes
```suggestion
    <Exec Command="&quot;$(IlcHostPackagePath)/native/src/libs/build-local.sh&quot; &quot;$(IlcHostPackagePath)/&quot; &quot;$(IntermediateOutputPath)&quot; System.Globalization.Native" Condition="'$(StaticICULinking)' == 'true'"/>

    <Exec Command="&quot;$(IlcHostPackagePath)/native/src/libs/build-local.sh&quot; &quot;$(IlcHostPackagePath)/&quot; &quot;$(IntermediateOutputPath)&quot; System.Security.Cryptography.Native" Condition="'$(StaticOpenSslLinking)' == 'true'"/>
```

**Comment by kant2002** [comment]
@jkotas is this wording, okay?

File: src/native/libs/build-local.sh [hunk 6]
```diff
@@ -4,24 +4,25 @@
 # The .NET Foundation licenses this file to you under the MIT license.
```
Comment (am11) [inline src/native/libs/build-local.sh:4]: Executable bit was dropped during the move: `chmod +x src/native/libs/build-local.sh`

Comment (kant2002) [reply src/native/libs/build-local.sh:4]: Thank you.

File: src/coreclr/nativeaot/docs/compiling.md [hunk 7]
```diff
@@ -90,3 +90,30 @@ Ubuntu (20.04+)
...
+    <PropertyGroup>
+      <StaticOpenSslLinking>true</StaticOpenSslLinking>
+    </PropertyGroup>
+```
+
+This feature is only supported on Linux. This feature is not supported when crosscompiling.
+
+### Prerequisites
+
+Alpine
```
Comment (am11) [inline src/coreclr/nativeaot/docs/compiling.md:116]:
````suggestion
Ubuntu
```
sudo apt-get install libssl-dev cmake
```

Alpine
````

File: src/native/libs/build-local.sh [hunk 8]
```diff
@@ -31,3 +32,5 @@ if [ -d "$SHIM_SOURCE_DIR" ]; then
         exit 1
     fi
 fi
+
+exit 0
```
Comment (am11) [inline src/native/libs/build-local.sh:36]:
```suggestion
```

**Comment by am11** [comment]
@kant2002, did you test it with code calling native API in System.Security.Cryptography and receive `no usable version of libssl was found` error? That error means we need to surround the dlopen calls with `#ifndef LOCAL_BUILD` in the initialization code.
https://github.com/dotnet/runtime/blob/d320fd90a2976b388169dd51b3732ad01bcb4266/src/native/libs/System.Security.Cryptography.Native/opensslshim.c#L55

(similar to https://github.com/dotnet/runtime/commit/7646e765eeb53f369338f9bbc6289ca88eafe1bd#diff-5c4725be11b8ef9741d6dba36be1244424dfcfc951ada9793862f002bcaee808)

**Comment by kant2002** [comment]
I did test in Docker container while making call to https website. https://github.com/kant2002/NativeAOTDocker/tree/main/OpenSslEmbedding

I did not test built product, but rather [patch](https://github.com/kant2002/NativeAOTDocker/blob/65c7bc49eb9b7cc1351539ac52483a4537baa172/OpenSslEmbedding/Dockerfile#L12-L15) existing ILC runtime nuget content.

Regarding ignoring opensslshim.c that file included if you build platformagnostic SSL, but I do not specify that define and things starts working.
https://github.com/dotnet/runtime/blob/d320fd90a2976b388169dd51b3732ad01bcb4266/src/native/libs/System.Security.Cryptography.Native/CMakeLists.txt#L58-L61

**Comment by jkotas** [comment]
@kant2002 Could you please resolve the conflict?

File: src/native/libs/build-local.sh (hunk 8 above)
Comment (am11) [inline src/native/libs/build-local.sh:36]: nit: rebase/merge conflicts
```suggestion
```

**Comment by davidfowl** [comment]
Is there an issue associated with this?

**Comment by kant2002** [comment]
@davidfowl  Technically no, because it start as experiment to create scratch console app with ICU in https://github.com/dotnet/runtime/issues/79498 and when I do not receive pushback on the scenario, I decide to make Web site on scratch container, like Go do it. Results is approximately 23MB for `dotnet new web` template using scratch, vs 33 MB using Chiselled Ubuntu.

If you interested, you can look at results of this experiment here https://github.com/kant2002/NativeAOTDocker

**Comment by davidfowl** [comment]
Can you file one?

**Review by jkotas — APPROVED** [review]
LGTM. Thank you!

**Pull request merged**